cpp/arcticdb/storage/memory_layout.hpp

```diff
@@ -0,0 +1,338 @@
+/* Copyright 2023 Man Group Operations Limited
+ *
+ * Use of this software is governed by the Business Source License 1.1 included in the file licenses/BSL.txt.
+ *
+ * As of the Change Date specified in that file, in accordance with the Business Source License, use of this software will be governed by the Apache License, version 2.0.
+ */
+#pragma once
+
+#include <cstdint>
+#include <array>
+#include "arcticdb/codec/magic_words.hpp"
+
+namespace arcticdb {
+
+#pragma pack(push)
+#pragma pack(1)
+
+constexpr size_t encoding_size = 6;
+// And extendable list of codecs supported by ArcticDB
+// N.B. this list is likely to change
+enum class Codec : uint16_t {
+    UNKNOWN = 0,
+    ZSTD,
+    PFOR,
+    LZ4,
+    PASS,
+    RLE,
+    FSST,
+    GORILLA_RLE,
+    CONSTANT
+};
+
+// Codecs form a discriminated union of same-sized objects
+// within the BlockCodec structure
+struct ZstdCodec {
+    static constexpr Codec type_ = Codec::ZSTD;
+
+    int32_t level_ = 0;
+    bool is_streaming_ = false;
+    uint8_t padding_ = 0;
+};
+
+static_assert(sizeof(ZstdCodec) == encoding_size);
+
+struct Lz4Codec {
+    static constexpr Codec type_ = Codec::LZ4;
+
+    int32_t acceleration_ = 1;
+    int16_t padding_ = 0;
+};
+
+static_assert(sizeof(Lz4Codec) == encoding_size);
+
+struct PassthroughCodec {
+    static constexpr Codec type_ = Codec::PASS;
+
+    uint32_t unused_ = 0;
+    uint16_t padding_ = 0;
```

Comment:
Something is going to go in the unused bit, just not completely sure what yet. Pfor is a special case 'cos it was speculatively in the protobufs, otherwise I'd just leave it out.